### DISCUSSION TITLE
Reflections on enum

### DISCUSSION THREAD

**Discussion by Davier** [bevyengine/bevy#1306] [Ideas]
The goal of this discussion is to design reflection on `enum` types. Ideally, `enum`s would at least support every case that the other types currently support.

## Nomenclature

I will try to use the same names as in `serde`.
```rust
// This enum has 4 "variants" (also called `EnumItem` in the [Rust Reference](https://doc.rust-lang.org/reference/items/enumerations.html))
enum TestEnum {
    A,                         // "unit" variant
    B(String),                 // "new type" variant
    C(usize, bool),            // "tuple" variant
    D { d1: usize, d2: bool }, // "struct" variant
}
```
Each variant is identified by a unique number called its discriminant. Discriminants are not indices, they may not be continuous and can even be set explicitly if all variants are unit variants (that is, in C-like `enum`s).

# Features

Let's start with a wish-list of features:
1. get the current variant, its name (e.g. `TestEnum::A`), and its discriminant
2. list the name and discriminant of all possible variants
3. support C-like `enum`s with explicit discriminants
4. extend `Reflect::reflect_ref()`
5. extend `#[derive(Reflect)]`
6. serialize and deserialize
7. set current value to another variant (constructing it with `Default` if supported)
8. `DynamicEnum`s

Does this look complete/reasonable?

# MVP

I propose as a first step to implement features 1 to 4. This means `enum`s could manually implement the `Enum` trait to get basic reflection. The other features can be built on it later, which gives us time to ~~bikeshed on~~ design them. I will present a possible API and associated design questions.

## `Enum` trait

Here is a first shot at the main trait that is the equivalent of the `Struct` trait.
```rust
pub trait Enum: Reflect {
    type Discriminant;
    /// Get the current variant
    fn variant(&self) -> EnumVariant<'_>;
    /// Get the name of the current variant
    fn variant_name(&self) -> &str;
    /// Get the discriminant of the current variant
    fn variant_discriminant(&self) -> Discriminant;
    /// Iterate all possible variants
    fn iter_variants(&self) -> impl Iterator<Item=Discriminant>;
    /// Get the number of possible variants
    fn variant_count(&self) -> usize;
    /// Get the specified variant if it's the correct one
    fn try_variant_at(&self, discriminant: Discriminant) -> Option<EnumVariant<'_>>;
    /// Get the name of the specified variant
    fn name_at(&self, discriminant: Discriminant) -> &str;
    /// Get the discriminant of a variant from its name
    fn get_discriminant_from_name(&self, name: &str) -> Result<Discriminant, _>;
    /// Get an integer representation of the specified discriminant
    fn get_discriminant_as_usize(discriminant: Discriminant) -> usize;
}
```
Is it missing any feature? Can you think of better names?

## `EnumVariant`

This is an `enum` that contains and describes the current variant. Following the [Serde data model](https://serde.rs/data-model.html), it defines 4 types of variants. It is conceptually similar to `enum ReflectRef` for `Reflect`. 
```rust
pub enum EnumVariant<'a> {
    Unit,
    NewType(&'a dyn Reflect),
    Tuple(&'a dyn Tuple),
    Struct(&'a dyn Struct),
}
```
An alternative would be to not define `NewType` and consider them as tuples of 1 element. It's unclear to me why serde makes the difference, the Rust Reference does not.

## `Discriminant`

This is the tricky part. It's possible to get the discriminant of any variant with `std::mem::discriminant(&TestEnum::A)`. It returns an opaque type: `std::mem::Discriminant<TestEnum>`. Since our `Discriminant` is an associated type, each implementation can set it to the correct `std::mem::Discriminant<T>`. That is enough for addressing since it implements `Eq` and `Hash` but in order to give an actual integer value (e.g. for serialization), we would like to convert that to an `usize` (`enum`s can never have a size larger than `usize`). For that we need to transmute it, however we have to avoid reading any uninitialized memory. A simple implementation of that looks like this:
```rust
// Result is unspecified but not unsound if T is not an enum
fn read_discriminant<T>(value: T) -> usize {
    use std::{intrinsics::transmute, mem::{discriminant, size_of, Discriminant}};
    let mut buffer = [0 as u8; size_of::<usize>()];
    let size = size_of::<Discriminant<T>>();
    let discriminant = discriminant(&value); // Warning: size_of_val(discriminant) != size_of::<Discriminant<T>>()
    let discriminant_ptr = &discriminant as *const Discriminant<T> as *const u8;
    for i in 0..size {
        buffer[i] = unsafe { discriminant_ptr.add(i).read() };
    }
    unsafe { transmute(buffer) }
}
```
There is probably a better way to do it, do tell me if you know it. Also, the compiler may be smart enough to optimize everything.

Other note: I used `usize` but the real discriminant's type might be signed. As far as I can tell there is no way to know from `std::mem::Discriminant<>` since it is opaque, but it can only happen in C-like `enum`s with the `#[repr(...)]` attribute and/or explicit discriminants, so it is possible to detect it in the derive macro. Then, we would have a second associated type that is either `usize` or `isize`, and transmute the discriminant to that type.

# Conclusion

This MVP seems to be implementable quickly when agreed on. I have not seriously investigated how to implement the other features, but I don't foresee any trouble. I will probably wait until we agree on this first part before detailing them here.

**Comment by Davier** [comment]
Now that I look back at the part about discriminant, I think I over-thought it. The fact that `size_of_val(discriminant(&value)) != size_of::<Discriminant<T>>()` should not matter when converting it to `usize` because this function is monomorphised (it did matter at first when I had a single type-erased discriminant, but I have dismissed that idea since).

**Reply by cart** [reply]
First: We can't make discriminant an associated type. We need all reflected enums to be the "same" type, otherwise we can't Box them generically: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9d18c8dad78e66db0a608f1f3934399e

I'm also pretty sure we shouldn't forcibly coerce discriminants into usize values.  Only c-like enums can be safely converted to integers. Everything else is an "opaque implementation detail". Couple that with the "typing" issue above, and I think rust-provided discriminants should not be part of the reflect api. People that need them can downcast to the real type and get the real discriminant that way.

I can think of a few options:

1.  Rely on names as the "discriminant". This requires no "new" concepts and plays nicely with the enum-related constraints rust imposes on us. The design already accounts for names, so this would really just mean removing discriminants from the api.
2. Create a new non-generic Discriminant type with a usize field. It should also probably include the TypeId so reflected enum discrimants can be compared across types. It would then be up to the implementer to determine what field corresponds to what index. The Reflect derive would just ascribe indices in declaration order for non-c-like-enums and would use direct int casts for c-like enums.
3. Same as 2 but instead of calling it Discriminant (and including the TypeId), just add something like `Enum::variant_usize -> usize`

(3) is my personal preference

**Reply by bjorn3** [reply]
> and would use direct int casts for c-like enums

what about

```rust
#[repr(u128)]
#[derive(Reflect)]
enum Foo {
    A = 0,
    B = -1i128 as u128,
}
```

?

**Reply by cart** [reply]
ooh fun fun. thanks for pointing that out. i guess we could:

1. Use u128 everywhere
2. Detect that case and fail the derive
3. Return a `AnyUsize { Usize(usize), U128(u128), ... }` enum type
4. Use the "order" int for every enum type.

Given that this is intended to be a general purpose reflection api, (2) feels unsatisfying. (1) only works until some new `#[repr(i128)]` enum comes along (or maybe that already exists?).  (3) would hamper ergonomics. (4) makes the integer information encoded in enums unavailable. So we would probably still want to add a new `-> Option<usize>` method / we still need to solve the problem.

@bjorn3 you have more context here than me, so I'm curious what your thoughts are.

**Reply by bjorn3** [reply]
`#[repr(i128)]` also exists as far as I know. I am leaning towards (4). It will keep working in the unlikely case that we get `u/i256` and I would expect the actual discriminant to not be useful for pretty much anything with `#[derive(Reflect)]` when using the concrete type wouldn't be possible.

**Reply by cart** [reply]
Yeah that seems reasonable to me. We might want to make it clear that it is the "declaration order". Something like `declaration_index() -> usize`. I could see users expecting `as_usize` to return the values defined here:

```rust
enum Foo {
    A = 4,
    B = 2,
} 
```

**Reply by Davier** [reply]
Thanks for the feedback!

>We can't make discriminant an associated type. We need all reflected enums to be the "same" type, otherwise we can't Box them generically

That's true. I will experiment with using an associated method instead.

>Create a new non-generic Discriminant type with a usize field. It should also probably include the TypeId so reflected enum discrimants can be compared across types.

That's what I referred to as "type-erased discriminant". Since it can't be generic I completely agree it should include a `TypeId` (regardless of whether we use discriminants or declaration order).

>I'm also pretty sure we shouldn't forcibly coerce discriminants into usize values. Only c-like enums can be safely converted to integers. Everything else is an "opaque implementation detail".

I'm not completely sure about that. The discriminant's type is opaque, but it is still specified in the Rust Reference as "an integer associated to [the enum] that is used to determine which variant it holds". I understand "integer" as "one of u8, u16, u32, u64 or their signed counterpart". I had also tested corner cases of `read_discriminant` with miri, and it didn't find any issue with non C-like enums (the values are the declaration order), but that only clears the current rustc implementation.
I suppose that if "integer" is used in a more general sense, an implementation could choose to use only some bits of the underlying type, masking the others in the Eq impl. That would break Eq in our type. It's certainly not what I observed on the two targets I tested (x86_64-unknown-linux-gnu and wasm32-unknown-unknown). The transmuted discriminant for `Option::Some(anything)` is always `1`. It also seems very unlikely to me since fieldless (i.e. C-like) enums do specify that the discriminant is "one of u8, u16, u32, u64 or their signed counterpart", equal to the declaration order (modulo the explicit assignments), so they would need to be treated separately.
We could decide that the current rustc implementation is the spec for now, add some assertions, and fix it if we are proven wrong in the future.

> what about #[repr(u128)]

I intentionally dismissed that case in my design. Using `repr(u128)` for enums requires an unstable, and most importantly incomplete feature. It has seen no progress since its introduction in 2018 (see https://github.com/rust-lang/rust/issues/56071), and it can cause the compiler to panic (see https://github.com/rust-lang/rust/issues/77457). It is apparently blocked by LLVM’s debuginfo API: https://github.com/rust-lang/rust/issues/35118#issuecomment-437683343. Since there doesn't seem to be any real use case for it, I doubt this feature will be complete and stable in a few years, if ever. The same should be true for the eventual `u256`. Does it seem reasonable to you to ignore them?

What we should probably support, however, is `i64` in addition to `u64`. So I suggest we use i128 as our representation of discriminants.

>Couple that with the "typing" issue above, and I think rust-provided discriminants should not be part of the reflect api. People that need them can downcast to the real type and get the real discriminant that way.

I originally made the discriminant part of the API because it was one of the given requirements for C-like enums and I didn't want to special-case them if possible. It seems like the natural way to do it to me. If after discussing it you prefer the declaration order, I will do that instead.

**Reply by Davier** [reply]
>People that need them can downcast to the real type and get the real discriminant that way.

>I would expect the actual discriminant to not be useful for pretty much anything with #[derive(Reflect)] when using the concrete type wouldn't be possible.

These are fair points. If the only use case is for debugging purpose, we could simply let all `Enum` impl defer to the `Debug` impl of their `std::mem::Discriminant`, which is simpler and completely safe.

**Reply by Davier** [reply]
Here is a more fleshed out definition:
```rust

pub trait Enum: Reflect {
    fn variant(&self) -> EnumVariant<'_>;
    fn variant_mut(&mut self) -> EnumVariantMut<'_>;
    fn variant_name(&self) -> &str;
    fn variant_discriminant(&self) -> ErasedDiscriminant;
    fn iter_variants(&self) -> Iter;
    // removed variant_count() since it is equivalent to iter_variants().count()
    fn try_variant_at(&self, discriminant: ErasedDiscriminant) -> Result<EnumVariant<'_>, EnumError>;
    fn try_variant_at_mut(&mut self, discriminant: ErasedDiscriminant) -> Result<EnumVariantMut<'_>, EnumError>;
    fn get_name_at(&self, discriminant: ErasedDiscriminant) -> Result<&str, EnumError>;
    fn get_discriminant_from_name(&self, name: &str) -> Option<ErasedDiscriminant>;
    // This method could be removed
    fn get_discriminant_as_integer(discriminant: ErasedDiscriminant) -> Result<i128, EnumError>;
}
pub struct ErasedDiscriminant {
    index: usize,
    type_id: TypeId,
}
impl ErasedDiscriminant {
    pub fn index(&self) -> usize {
        self.index
    }
    pub fn enum_type_id(&self) -> TypeId {
        self.type_id
    }
}
pub struct Iter {
    current_index: usize,
    len: usize,
    type_id: TypeId,
}
impl Iterator for Iter{
    type Item=ErasedDiscriminant;
    fn next(&mut self) -> Option<Self::Item> {
        if self.current_index == self.len {
            return None;
        }
        let item = ErasedDiscriminant {
            index: self.current_index,
            type_id: self.type_id,
        };
        self.current_index += 1;
        Some(item)
    }
}
use thiserror::Error;
#[derive(Error, Debug)]
pub enum EnumError {
    #[error("tried to access the incorrect variant")]
    InvalidVariant,
    #[error("tried to access a variant from a different enum type")]
    InvalidType,
}
pub enum EnumVariant<'a> {
    Unit,
    NewType(&'a dyn Reflect),
    Tuple(&'a dyn Tuple),
    Struct(&'a dyn Struct),
}
pub enum EnumVariantMut<'a> {
    Unit,
    NewType(&'a mut dyn Reflect),
    Tuple(&'a mut dyn Tuple),
    Struct(&'a mut dyn Struct),
}
```
`ErasedDiscriminant` is not a great name. Is `Discriminant` better, even with the name collision with `std::mem::Discriminant`?

`Enum::get_discriminant_as_integer()` could be implemented using `as` for C-like enums and returning `ErasedDiscriminant::index` otherwise.

**Reply by cart** [reply]
I'm thinking try_variant_at and try_variant_at_mut could be removed in favor of:

```rust
if value.variant_discriminant() == discriminant {
  let variant = value.variant(); // or alternatively variant_mut()
}
```

iter_variants() is probably the wrong name, given that we aren't returning EnumVariant. it should probably be `iter_discriminants()` in its current form

I think I like `get_discriminant_name(discriminant)` over `get_name_at`. I'm not super happy with `Struct::name_at` either (which im sure is what motivated your choice). It should probably be `get_index_name()` or something.


I think for now I'm fine with leaving enum integer casts as an exercise for the user. They aren't common across enum types and I don't want to conflate "integer casts" with "definition order" semantically. "castability" is something enforced / determined by the compiler, so it makes sense to me to just force users to downcast when that behavior is required.

I'm also not sure we "need" type agnostic discriminants. Reflect already gives us the ability to retrieve the TypeId, which can be added to a comparison when necessary.

We could consider returning something like VariantInfo, which would contain the index and the static string name. This would allow us to replace `variant_name`, `variant_discriminant`, and `iter_variants` with a `variant_info` and `iter_variant_info`

**Reply by Davier** [reply]
>I'm also not sure we "need" type agnostic discriminants. Reflect already gives us the ability to retrieve the TypeId, which can be added to a comparison when necessary.

Since the type validation is no longer at compile time, I agree.

Here is the updated version, getting rid completely of the discriminant:
```rust
pub trait Enum: Reflect {
    fn variant(&self) -> EnumVariant<'_>;
    fn variant_mut(&mut self) -> EnumVariantMut<'_>;
    fn variant_info(&self) -> VariantInfo<'_>;
    fn iter_variants_info(&self) -> VariantInfoIter<'_>;
    fn get_index_name(&self, index: usize) -> Option<&str>;
    fn get_index_from_name(&self, name: &str) -> Option<usize>;
}
pub struct VariantInfo<'a> {
    pub index: usize,
    pub name: &'a str,
}
```

I am moving forward, trying to manually implement the trait. For now everything works except for the two "interesting" cases: `EnumVariant::Tuple` and `EnumVariant::Struct`. The issue is that since the variants don't have a type, so we can't return references to it. It's also not possible AFAIK to declare a type that has the same layout as a variant (otherwise we could just transmute the reference to that).
The solution is probably to return some sort of smart pointer with custom implementations of `Reflect` and `Struct`/`Tuple`, however I'm stuck on that. It looks like structs that have a lifetime parameter (due to the inner reference to the enum) cannot implement Reflect, because  `Reflect: Any` and `Any: 'static`. Would you have any idea on how to deal with that? 

Alternative solutions could be:
1. don't reflect these variants. If an enum needs reflection on variants containing data, require that it uses a `NewType` instead (obviously not great, but it already works)
2. declare equivalent types of these variants, and copy the variant's fields into them (not compatible with `variant_mut`)
3. similar to 2., but using `DynamicStruct` and `DynamicTuple` instead of declaring the equivalent types (same issue as 2.)
4. it might be possible inside the derive macro for the enum to declare a type that has the same layout as a variant?

**Reply by bjorn3** [reply]
Could you store a reference to the whole enum and then when accessing fields of the selected enum variant match on the enum?

**Reply by Davier** [reply]
> Could you store a reference to the whole enum and then when accessing fields of the selected enum variant match on the enum?

If I understand you correctly, that's what I'm trying to do with the "smart pointer" solution. But in order to actually expose the fields of the variant to the user, it would need to implement the `Struct` trait, which requires a `'static` lifetime, which prevents me from storing the reference the the enum.

**Reply by Davier** [reply]
This is what I am trying to do (I think taking a ref to the enum would have the same issues as taking refs to the fields like here):
```rust
pub enum EnumVariant<'a> {
    Unit,
    NewType(&'a dyn Reflect),
    Tuple(Box<dyn Tuple + 'a>), // error: lifetime parameter must outlive the static lifetime
    Struct(Box<dyn Struct + 'a>), // error: lifetime parameter must outlive the static lifetime
}
#[derive(Reflect, Clone)]
pub enum TestEnum {
    A,
    B(String),
    C(usize, bool),
    D { d1: usize, d2: bool },
}
pub struct TestEnumCRef<'a>(&'a usize, &'a bool);
impl<'a> Reflect for TestEnumCRef<'a> {/* TODO */} // error: lifetime parameter must outlive the static lifetime
impl<'a> Tuple for TestEnumCRef<'a> {/* TODO */} // error: lifetime parameter must outlive the static lifetime
pub struct TestEnumDRef<'a> {
    d1: &'a usize,
    d2: &'a bool,
}
impl<'a> Reflect for TestEnumDRef<'a> {/* TODO */} // error: lifetime parameter must outlive the static lifetime
impl<'a> Struct for TestEnumDRef<'a> {/* TODO */} // error: lifetime parameter must outlive the static lifetime
impl Enum for TestEnum {
    fn variant(&self) -> EnumVariant<'_> {
        match self {
            TestEnum::A => EnumVariant::Unit,
            TestEnum::B(new_type) => EnumVariant::NewType(new_type as &dyn Reflect),
            TestEnum::C(c1, c2) => {
                EnumVariant::Tuple(Box::new(TestEnumCRef(c1, c2)) as Box<dyn Tuple>)
            }
            TestEnum::D { d1, d2 } => {
                EnumVariant::Struct(Box::new(TestEnumDRef { d1, d2 }) as Box<dyn Struct>)
            }
        }
    }
    // [snip]
}
```

**Reply by Davier** [reply]
I believe I have a solution :-)
It works without storing a reference to the enum in another struct. The reference to the enum is transmuted to a reference to a transparent type containing it (I think this is always sound?), and that type implements `Reflect` and `Tuple` to do the dispatch. So it's sort of solution 4.
```rust
#[repr(transparent)]
pub struct TestEnumCWrapper (TestEnum);
impl Reflect for TestEnumCWrapper {/* TODO */}
impl Tuple for TestEnumCWrapper {
    fn field(&self, index: usize) -> Option<&dyn Reflect> {
        let (c1, c2) = match &self.0 {
            TestEnum::A => {unreachable!()}
            TestEnum::B(_) => {unreachable!()}
            TestEnum::C(c1, c2) => {(c1, c2)}
        };
        match index {
            0 => Some(c1),
            1 => Some(c2),
            _ => None,
        } 
    }
    // [snip]
}
impl Enum for TestEnum {
    fn variant(&self) -> EnumVariant<'_> {
        match self {
            TestEnum::A => EnumVariant::Unit,
            TestEnum::B(new_type) => EnumVariant::NewType(new_type as &dyn Reflect),
            TestEnum::C(..) => {
                let wrapper_ref = unsafe{std::mem::transmute::<&Self, &TestEnumCWrapper>(self)};
                EnumVariant::Tuple(wrapper_ref as &dyn Tuple)
            }
            TestEnum::D { d1, d2 } => {
                todo!()
            }
        }
    }
    // [snip]
}
```

**Reply by cart** [reply]
Haha I've learned the hard way that transmute is basically never the right solution to a problem. I'll start poking around.

**Reply by bjorn3** [reply]
In this case the transmute should be fine. It  is sinply transmuting from a `TestEnum` reference to a reference to a `TestEnum` wrapper that is guaranteed to have the same layout.

**Reply by Davier** [reply]
#1347 now implements the MVP, and the derive macro. The derive macro is not ready for a proper review yet, but it helps a lot to try out the trait.
